### PR TITLE
Chapter 12 applicative

### PR DESCRIPTION
**Pull request by gpoirier**

### CODE REVIEW

File: exercises/src/main/scala/fpinscala/applicative/Applicative.scala [hunk 1]
```diff
@@ -4,38 +4,88 @@ package applicative
...
 
-  def product[G[_]](G: Applicative[G]): Applicative[({type f[x] = (F[x], G[x])})#f] = ???
 
-  def compose[G[_]](G: Applicative[G]): Applicative[({type f[x] = F[G[x]]})#f] = ???
+  def replicateM[A](n: Int, fa: F[A]): F[List[A]] =
+    sequence(List.fill(n)(fa))
+
+  def factor[A,B](fa: F[A], fb: F[B]): F[(A,B)] = map2(fa, fb)((_, _))
+
+  def product[G[_]](G: Applicative[G]): Applicative[({type f[x] = (F[x], G[x])})#f] = {
```
Comment (MasseGuillaume) [inline exercises/src/main/scala/fpinscala/applicative/Applicative.scala:57]: https://github.com/non/kind-projector

File: exercises/src/test/scala/fpinscala/applicative/ApplicativeSpec.scala [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+class ApplicativeSpec extends Specification {
+
+  "streamApplicative" >> {
+    "sequence" >> {
+
+      skipped("sequences cause a endless loop because the stream is infinite")
+
+      val A = streamApplicative
+      val streams = (1 to 10).map(i => A.unit(i)).toList
```
Comment (MasseGuillaume) [inline exercises/src/test/scala/fpinscala/applicative/ApplicativeSpec.scala:19]: does not look infinite

Comment (gpoirier) [inline exercises/src/test/scala/fpinscala/applicative/ApplicativeSpec.scala:19]: The applicative `A` produces an infinite stream.

File: exercises/src/test/scala/fpinscala/applicative/ApplicativeSpec.scala [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        else
+          Failure(s"Phone doesn't match expected format (XXX-XXX-XXXX): $text")
+
+      val form = Applicative.validationApplicative.map3(
+        validateName("Guillaume"),
+        validateBirthDate("1980/04/16"),
+        validatePhone("1-514-111-1111")
+      ) {
+        WebForm.apply
+      }
```
Comment (MasseGuillaume) [inline exercises/src/test/scala/fpinscala/applicative/ApplicativeSpec.scala:103]:
``` scala
import scalaz._
import Scalaz._
import scala.util.control.NonFatal
import java.util.Date

val fmt = new java.text.SimpleDateFormat("yyyy-MM-dd")

type Val[T] = Validation[NonEmptyList[String], T]

def validateName(text: String): Val[String] =
    if (text == "Guillaume") text.successNel
    else s"Unexpected name: $text".failureNel

def validateBirthDate(text: String): Val[Date] =
    try {
    fmt.parse(text).successNel
  } catch {
    case NonFatal(e) => e.getMessage.failureNel
  }

def validatePhone(text: String): Val[String] =
    if (text.matches(raw"\d{3}-\d{3}-\d{4}"))
        text.successNel
    else
        s"Phone doesn't match expected format (XXX-XXX-XXXX): $text".failureNel


case class Person(name: String, bday: Date, phone: String)

def validate(name: String, bday: String, phone: String) = {
  ( validateName(name) |@| 
    validateBirthDate(bday) |@| 
    validatePhone(phone)
  )(Person)
}

validate("Guiaume", "5-5-", "a")
// Failure(NonEmptyList(Unexpected name: Guiaume, Unparseable date: "5-5-", Phone doesn't match expected format (XXX-XXX-XXXX): a))
validate("Guillaume", "5-5-5", "514-444-4444")
// Success(Person(Guillaume,Tue May 05 00:00:00 EST 5,514-444-4444))
```

**Pull request merged**